### PR TITLE
Fallback using the `CultureInfo.Parent` instead of using hyphen

### PR DESCRIPTION
**Pull request by hishamco**
Nowadays `RequestLocalizationMiddleware` is using hyphen for fallback to the parent culture(s) which is different behavior than other places where we are using `CultureInfo.Parent`

The issue is might occur with Chinsese cultures for instance `zh-Hans-CN`, this will fallback into `zh-Hans` which is totally wrong, the best approach to get the parent culture using `CultureInfo.Parent` this way we will get `zh-CHS`, `zh-Hans`, `zh` respectively

### CODE REVIEW

**Comment by hishamco** [comment]
I might add a unit tests if we agreed about this PR

**Comment by hishamco** [comment]
Seems something weird happen on evey version of .NET Core, especially with Chinese cultures :)

**Comment by hishamco** [comment]
FYI the PR will make no difference in .NET 7 while the fallback: `zh-Hans-CN` -> `zh-Hans` -> `zh`,  while in .NET Framework 4.7 and Roslyn 4.4 `zh-CHS` should be there on the fallback hierarchy

**Comment by adityamandaleeka** [comment]
@tarekgh Can you please take a quick look to verify this approach?

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs
Comment (tarekgh) [inline src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:163]: It is better to short circuit exiting if the culture is invariant. I mean when `cultureName.ToString()` is empty string.

**Review by tarekgh — APPROVED** [review]

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs
Comment (adityamandaleeka) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:163]: @hishamco Can you please make this change? Other than that, this PR should be good to go.

**Comment by mitchdenny** [comment]
I've created a new PR to replace this one with a unit test to cover this scenario (to make sure we ware walking the parents):

https://github.com/dotnet/aspnetcore/pull/47365

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs
Comment (mitchdenny) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:163]: See linked PR which includes @tarekgh's feedback. Once we are comfortable that my new PR is good we can probably just close this one down.

Thanks for submitting the code @hishamco - I basically just lifted your change and added a unit test.

**Comment by hishamco** [comment]
@mitchdenny I can write a unit tests directly on this PR if someone askes for, also I notice there's less activity on this PR since I pushed

File: src/Middleware/Localization/test/UnitTests/RequestLocalizationMiddlewareTest.cs
Comment (mitchdenny) [inline src/Middleware/Localization/test/UnitTests/RequestLocalizationMiddlewareTest.cs:49]: This assertion doesn't really prove that the code is walking the parent hierarchy vs. chopping up the string. Even though you have the ```zh-TW``` test case in here, what you really want to assert is that it can find ```zh-Hant``` as a supported culture (implies that it is in the list of supported cultures).

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs
Comment (tarekgh) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:163]: The root of all cultures is the Invariant culture which has the name as empty string. If you encounter it that means you reached the super root. 

> if the string is empty, should we return CultureInfo.InvariantCulture (open question).

Is the Invariant culture one of the supported cultures in the list `supportedCultures`? if yes, then the answer of your question is yes too.

Comment (mitchdenny) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:163]: @hishamco I've had a closer look at the code and I think returning null when cultureName is an empty string is the desirable approach. If we don't do this we needlessly spin down the max-depth when traversing the parent property.

**Review by mitchdenny — CHANGES_REQUESTED** [review]
Make sure we short circuit when `cultureName == ""`

File: src/Middleware/Localization/test/UnitTests/RequestLocalizationMiddlewareTest.cs
Comment (hishamco) [reply src/Middleware/Localization/test/UnitTests/RequestLocalizationMiddlewareTest.cs:49]: > what you really want to assert is that it can find zh-Hant as a supported culture (implies that it is in the list of supported cultures).

This should be easy if the `zh-Hant` is supported culture

**Comment by hishamco** [comment]
> Make sure we short circuit when cultureName == ""

I might disagree, because when the culture is `InvariantCulture` we SHOULD return it instead of returning a `null`

**Comment by mitchdenny** [comment]
@hishamco if you look at the existing implementations of `GetCultureInfo` through that method, none of them return `CultureInfo.CultureInvariant`. It makes sense because the intention of this method isn't to return the nearest common culture from all cultures, its to return the nearest common culture from the list of supported cultures.

Its the responsibility of the caller to determine what happens if such a match couldn't be found. For example, if I I said that the supported cultures in my app are "fr" and "en", and the user agent submits "zh-TW" then the correct answer is that there was no common culture - not invariant culture.

**Comment by hishamco** [comment]
Let us complete the discussion on the original PR instead

**Comment by hishamco** [comment]
https://github.com/dotnet/aspnetcore/blob/94e664ee9a7e4f12930ca5ef8f26e59fe168b44a/src/Middleware/Localization/src/RequestLocalizationMiddleware.cs#L170-L191

Can you tell me why we didn't check against invariant culture on the current implementation (before my PR)?

**Comment by mitchdenny** [comment]
I can't comment on that. I didn't write the code. However changing from processing the string to walking the parent property does make it more like they well soon down to max depth (when an unsupported culture is supported).

The short circuit just optimises that scenario out.

**Comment by hishamco** [comment]
I suggest opening another issue, after merging this PR describing the short circuit that you're suggesting, then let's involve the guys who are interested on the localization area

**Comment by mitchdenny** [comment]
> I suggest opening another issue, after merging this PR describing the short circuit that you're suggesting, then let's involve the guys who are interested on the localization area

Appreciate the suggestion. However, because this PR increases the number of recursions that are going to occur and it's in the hot path for each and every request, we are going to need the short circuit in there before we can approve and merge this PR. If you make the change then I can approve and merge, alternatively I'll wait for someone on the team to approve my PR and I'll merge it (then we can close this PR).

**Comment by hishamco** [comment]
No problem I can make the change on this PR, but one more last question, when the culture will be empty?

**Comment by hishamco** [comment]
No problem I can make the change on this PR, but one more last question, when the culture will be empty?

**Review by mitchdenny — APPROVED** [review]

**Pull request merged**
